util/containers_linux.go

```diff
@@ -0,0 +1,75 @@
+// +build linux
+
+package util
+
+import (
+	"errors"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/cache"
+	"github.com/DataDog/datadog-agent/pkg/util/containers"
+	"github.com/DataDog/datadog-agent/pkg/util/containers/collectors"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var detector *collectors.Detector
+var containerCacheDuration = 10 * time.Second
+
+// SetContainerSource allows config to force a single container source
+func SetContainerSource(name string) {
+	detector = collectors.NewDetector(name)
+}
+
+// GetContainers returns containers found on the machine, autodetecting
+// the best backend from available sources
+func GetContainers() ([]*containers.Container, error) {
+	// Detect source
+	if detector == nil {
+		detector = collectors.NewDetector("")
+	}
+	l, name, err := detector.GetPreferred()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get containers from cache and update metrics
+	cacheKey := cache.BuildAgentKey("containers", name)
+	cached, hit := cache.Cache.Get(cacheKey)
+	if hit {
+		containers, ok := cached.([]*containers.Container)
+		if !ok {
+			log.Errorf("Invalid container list cache format, forcing a cache miss")
+			hit = false
+		} else {
+			err := l.UpdateMetrics(containers)
+			log.Infof("Got %d containers from cache", len(containers))
+			return containers, err
+		}
+	}
+	// If cache empty/expired, get a new container list
+	if !hit {
+		containers, err := l.List()
+		if err != nil {
+			return nil, err
+		}
+		cache.Cache.Set(cacheKey, containers, containerCacheDuration)
+		log.Infof("Got %d containers from source %s", len(containers), name)
+		return containers, nil
+	}
+	return nil, errors.New("")
```

Comment:
I could imagine some code change that has us hitting this empty error. should we just drop the `if !hit {` and follow the normal path? or just an `else`?